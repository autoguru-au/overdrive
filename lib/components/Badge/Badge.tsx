import cx from 'clsx';
import React, { FunctionComponent, memo } from 'react';
import styles from './style.scss';

export interface IProps {
	colour?: EColour;
	className?: string;
	invert?: boolean;
	label: string;
}

export enum EColour {
	Default = 'default',
	Success = 'success',
	Warning = 'warning',
	Danger = 'danger',
}

const cssColourMap = new Map([
	[EColour.Default, styles.colourDefault],
	[EColour.Success, styles.colourSuccess],
	[EColour.Warning, styles.colourWarning],
	[EColour.Danger, styles.colourDanger],
]);

const BadgeComponent: FunctionComponent<IProps> = ({
	label,
	colour = EColour.Default,
	invert = false,
	className = '',
}) => {
	const cls = cx([className, styles.root, cssColourMap.get(colour)], {
		[styles.inverted]: invert,
	});

	return <span className={cls}>{label}</span>;
};

export const Badge = memo(BadgeComponent);
