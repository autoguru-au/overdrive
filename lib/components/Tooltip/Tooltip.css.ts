import { style } from '@vanilla-extract/css';

import { responsiveStyle } from '../../utils/responsiveStyle';

export const root = style({
	maxWidth: '250px',
	...responsiveStyle({
		tablet: {
			maxWidth: '450px',
		},
	}),
});
