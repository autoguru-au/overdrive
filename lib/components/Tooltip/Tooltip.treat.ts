import { style } from 'treat';

export const root = style({
	overflow: 'hidden',
	width: '100%',
	maxWidth: '256px',
	userSelect: 'none',
	pointerEvents: 'none',
});
