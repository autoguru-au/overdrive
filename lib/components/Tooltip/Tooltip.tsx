import React, {
	Children,
	cloneElement,
	FunctionComponent,
	ReactElement,
	useEffect,
	useRef,
	useState,
} from 'react';
import { EPositionerAlignment, Positioner } from '../Positioner';
import { Text } from '../Typography';

import styles from './style.scss';

interface Props {
	label: string;
	alignment?: EPositionerAlignment;
	children: ReactElement;
}

export const Tooltip: FunctionComponent<Props> = ({
	alignment = EPositionerAlignment.RIGHT,
	label,
	children,
}) => {
	const [isOpen, setIsOpen] = useState(false);
	const childRef = useRef<HTMLDivElement>();
	const triggerRef = useRef<HTMLElement>();

	const enterHandlers = ['mouseenter', 'touchenter'];
	const leaveHandlers = ['mouseleave', 'touchleave'];

	useEffect(() => {
		let leaveTimer;
		let current = true;

		if (triggerRef.current) {
			const close = () => {
				if (current) {
					setIsOpen(false);
				}
			};

			const open = () => {
				if (current) {
					setIsOpen(true);
				}
			};

			const enterHandler = () => {
				if (leaveTimer) {
					clearTimeout(leaveTimer);
				}

				open();
			};

			const leaveHandler = () => {
				leaveTimer = setTimeout(() => {
					close();
				}, 1e3 / 2);
			};

			enterHandlers.forEach(handler =>
				triggerRef.current.addEventListener(handler, enterHandler, {
					passive: true,
				}),
			);
			leaveHandlers.forEach(handler =>
				triggerRef.current.addEventListener(handler, leaveHandler, {
					passive: true,
				}),
			);

			return () => {
				current = false;

				if (triggerRef.current) {
					enterHandlers.forEach(handler =>
						triggerRef.current.removeEventListener(
							handler,
							enterHandler,
						),
					);
					leaveHandlers.forEach(handler =>
						triggerRef.current.removeEventListener(
							handler,
							leaveHandler,
						),
					);
				}
			};
		}

		return () => {
			current = false;
		};
	}, [enterHandlers, leaveHandlers, triggerRef]);

	return (
		<>
			{Children.only(children) &&
				cloneElement(children, {
					ref: triggerRef,
				})}
			<Positioner
				triggerRef={triggerRef}
				alignment={alignment}
				isOpen={isOpen}>
				<div ref={childRef} className={styles.root}>
					<Text white>{label}</Text>
				</div>
			</Positioner>
		</>
	);
};
