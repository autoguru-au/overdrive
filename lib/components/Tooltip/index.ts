export { Tooltip, type TooltipProps } from './Tooltip';
