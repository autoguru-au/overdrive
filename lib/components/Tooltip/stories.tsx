import { select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { Tooltip } from '.';
import { EPositionerAlignment } from '../Positioner';

const alignmentPicker = () =>
	select('Alignment', EPositionerAlignment, EPositionerAlignment.RIGHT);

storiesOf('Components|Tooltip', module).add('default', () => (
	<Tooltip alignment={alignmentPicker()} label="Im the tooltip body">
		<div>Im the tooltip trigger</div>
	</Tooltip>
));
