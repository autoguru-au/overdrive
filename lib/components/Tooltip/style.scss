@import '../../theme/tokens/facade';

.root {
	overflow: hidden;
	padding: var(--global--space--4);
	width: 100%;
	max-width: 256px;
	color: white;
	border: 1px solid getColour(gray, 800);
	border-radius: var(--global--space--1);
	background-color: getColour(gray, 900);
	box-shadow: var(--global--elevationBoxShadow--4);
	user-select: none;
	pointer-events: none;
}
