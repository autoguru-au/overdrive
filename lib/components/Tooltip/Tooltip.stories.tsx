import { Meta, StoryObj } from '@storybook/react-vite';
import React, { type ComponentProps } from 'react';

import { EAlignment } from '../Positioner/alignment';
import { Text } from '../Text/Text';

import { Tooltip } from './Tooltip';

const sizeScale: ComponentProps<typeof Text>['size'][] = ['3', '4'];

const meta = {
	title: 'Components/Tooltip',
	component: Tooltip,
	decorators: [
		(Story) => (
			<div style={{ marginLeft: 100, marginTop: 100 }}>
				<Story />
			</div>
		),
	],
	parameters: {
		chromatic: { disable: true },
	},
	argTypes: {
		alignment: {
			options: Object.values(EAlignment),
			defaultValue: EAlignment.RIGHT,
			control: {
				type: 'select',
			},
		},
		size: {
			options: sizeScale,
			defaultValue: void 0,
			control: {
				type: 'select',
			},
		},
		label: {
			defaultValue: '',
		},
		closeAfter: {
			defaultValue: EAlignment.RIGHT,
			control: {
				type: 'number',
			},
		},
	},
} satisfies Meta<typeof Tooltip>;

export default meta;

type Story = StoryObj<typeof Tooltip>;

export const Standard: Story = {
	args: {
		label: "I'm the tooltip body",
		closeAfter: undefined,
		children: (
			<div style={{ display: 'inline' }}>Im the tooltip trigger</div>
		),
	},
};

export const WithAutoClose: Story = {
	args: {
		label: 'I will automatically close after 5 seconds',
		closeAfter: 5e3,
		children: (
			<div style={{ display: 'inline' }}>Im the tooltip trigger</div>
		),
	},
};

export const WithLongText: Story = {
	args: {
		label: 'Far far away, behind the word mountains, far from the countries Vokalia and Consonantia, there live the blind texts.',
		alignment: EAlignment.BOTTOM,
		children: (
			<div style={{ display: 'inline' }}>Im the tooltip trigger</div>
		),
	},
};

export const WithLargeTextSize: Story = {
	args: {
		label: 'Far far away, behind the word mountains, far from the countries Vokalia and Consonantia, there live the blind texts.',
		alignment: EAlignment.BOTTOM,
		size: 'large',
		children: (
			<div style={{ display: 'inline' }}>Im the tooltip trigger</div>
		),
	},
};
