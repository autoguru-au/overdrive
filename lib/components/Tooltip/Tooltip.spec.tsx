import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { vi } from 'vitest';

import { Badge } from '../Badge';
import { Button } from '../Button';
import { Text } from '../Text';
import { TextInput } from '../TextInput';

import { Tooltip } from './Tooltip';

describe('<Tooltip />', () => {
	const tooltipContentText = 'tooltip content';

	it('should not throw', () => {
		expect(() => {
			render(
				<Tooltip label={tooltipContentText}>
					<div>trigger</div>
				</Tooltip>,
			);
		}).not.toThrow();
	});

	it('should have body when open', () => {
		const { container, getByText } = render(
			<Tooltip label={tooltipContentText}>
				<div>trigger</div>
			</Tooltip>,
		);

		fireEvent.mouseEnter(getByText('trigger'));

		expect(container.parentNode).toHaveTextContent(tooltipContentText);
	});

	it('should match snapshot when closed', () => {
		const { baseElement } = render(
			<Tooltip label={tooltipContentText} wrapper>
				<div>trigger</div>
			</Tooltip>,
		);

		expect(baseElement).toMatchSnapshot();
	});

	it('should match snapshot when opened', () => {
		const { baseElement, getByText } = render(
			<Tooltip label={tooltipContentText} wrapper>
				<div>trigger</div>
			</Tooltip>,
		);

		fireEvent.mouseEnter(getByText('trigger'));

		expect(baseElement).toMatchSnapshot();
	});

	describe('with string children', () => {
		it('should work with string content when using wrapper', () => {
			const { container, getByText } = render(
				<Tooltip label={tooltipContentText} wrapper>
					Simple text content
				</Tooltip>,
			);

			expect(container).toHaveTextContent('Simple text content');

			fireEvent.mouseEnter(getByText('Simple text content'));
			expect(container.parentNode).toHaveTextContent('tooltip content');
		});

		it('should fail with string content when not using wrapper', () => {
			// This should throw an invariant error since string children require wrapper
			expect(() => {
				render(
					<Tooltip label="tooltip content">
						Simple text content
					</Tooltip>,
				);
			}).toThrow();
		});
	});

	describe('with Text component children', () => {
		it('should work with Text component', () => {
			const { container } = render(
				<Tooltip label={tooltipContentText}>
					<Text>Text component content</Text>
				</Tooltip>,
			);

			expect(container).toHaveTextContent('Text component content');

			// Tooltip should wrap the Text component in a span
			const wrapper = container.querySelector('span');
			expect(wrapper).toBeInTheDocument();

			if (wrapper) fireEvent.mouseEnter(wrapper);
			expect(container.parentNode).toHaveTextContent(tooltipContentText);
		});
	});

	describe('with Badge component children', () => {
		it('should work with Badge component', async () => {
			const { container } = render(
				<Tooltip label={tooltipContentText} wrapper>
					<Badge label="Badge content" />
				</Tooltip>,
			);

			// Find the Badge element directly and trigger hover
			const badge = container.querySelector(
				'[data-od-component="badge"]',
			) as HTMLElement;

			expect(badge).toBeInTheDocument();
			expect(container).toHaveTextContent('Badge content');

			fireEvent.mouseEnter(badge);
			expect(document.body).toHaveTextContent(tooltipContentText);
		});
	});

	describe('multiple children support', () => {
		it('should work with multiple children when using wrapper', () => {
			const { container } = render(
				<Tooltip label={tooltipContentText} wrapper>
					<div>First child</div>
					<div>Second child</div>
				</Tooltip>,
			);

			expect(container).toHaveTextContent('First child');
			expect(container).toHaveTextContent('Second child');

			// Multiple children should be wrapped in a span
			const wrapper = container.querySelector('span');
			expect(wrapper).toBeInTheDocument();
		});

		it('should show tooltip when hovering over multiple children wrapper', () => {
			const { container } = render(
				<Tooltip label={tooltipContentText} wrapper>
					<div>First child</div>
					<div>Second child</div>
				</Tooltip>,
			);

			const wrapper = container.querySelector('span');
			if (wrapper) fireEvent.mouseEnter(wrapper);

			expect(container.parentNode).toHaveTextContent(tooltipContentText);
		});

		it('should fail with multiple children when not using wrapper', () => {
			expect(() => {
				render(
					<Tooltip label="Test">
						<div>First child</div>
						<div>Second child</div>
					</Tooltip>,
				);
			}).toThrow();
		});

		it('should work with React Fragment children when using wrapper', () => {
			const { container } = render(
				<Tooltip label={tooltipContentText} wrapper>
					<>
						<div>Fragment child 1</div>
						<div>Fragment child 2</div>
					</>
				</Tooltip>,
			);

			expect(container).toHaveTextContent('Fragment child 1');
			expect(container).toHaveTextContent('Fragment child 2');

			// Fragment children should be wrapped in a span
			const wrapper = container.querySelector('span');
			expect(wrapper).toBeInTheDocument();

			if (wrapper) fireEvent.mouseEnter(wrapper);

			expect(container.parentNode).toHaveTextContent(tooltipContentText);
		});
	});

	describe('controlled state', () => {
		it('should show tooltip when isOpen is true', () => {
			const { container } = render(
				<Tooltip label="controlled tooltip" isOpen={true}>
					<div>trigger</div>
				</Tooltip>,
			);

			expect(container.parentNode).toHaveTextContent(
				'controlled tooltip',
			);
		});

		it('should not show tooltip when isOpen is false', () => {
			const { container } = render(
				<Tooltip label="controlled tooltip" isOpen={false}>
					<div>trigger</div>
				</Tooltip>,
			);

			expect(container.parentNode).not.toHaveTextContent(
				'controlled tooltip',
			);
		});
	});

	describe('no label handling', () => {
		it('should not render tooltip when label is empty', () => {
			const { container } = render(
				<Tooltip label="">
					<div>trigger</div>
				</Tooltip>,
			);

			// Should return children as-is without any wrapper
			expect(container).toHaveTextContent('trigger');
			expect(container.querySelector('span')).not.toBeInTheDocument();
		});
	});
});

describe('<Tooltip /> with interactive child', () => {
	it('should not throw with interactive components', () => {
		expect(() => {
			render(
				<Tooltip label="Interactive tooltip">
					<Button>Click me</Button>
				</Tooltip>,
			);
		}).not.toThrow();
	});

	describe('with Button component', () => {
		it('should work with Button component', () => {
			const { getByRole } = render(
				<Tooltip label="Button tooltip">
					<Button>Interactive Button</Button>
				</Tooltip>,
			);

			const button = getByRole('button');
			expect(button).toHaveTextContent('Interactive Button');

			fireEvent.mouseEnter(button);
			// Use a more reliable way to check for tooltip content
			expect(document.body).toHaveTextContent('Button tooltip');
		});

		it('should preserve button functionality', () => {
			const handleClick = vi.fn();
			const { getByRole } = render(
				<Tooltip label="Clickable button tooltip">
					<Button onClick={handleClick}>Clickable Button</Button>
				</Tooltip>,
			);

			const button = getByRole('button');
			fireEvent.click(button);
			expect(handleClick).toHaveBeenCalledTimes(1);
		});
	});

	describe('with TextInput component', () => {
		it('should work with TextInput component', () => {
			const { container, getByRole } = render(
				<Tooltip label="Input tooltip">
					<TextInput name="test-input" placeholder="Enter text" />
				</Tooltip>,
			);

			const input = getByRole('textbox');

			fireEvent.mouseEnter(input);
			expect(container.parentNode).toHaveTextContent('Input tooltip');
		});

		it('should preserve input functionality', () => {
			const handleChange = vi.fn();
			const { getByRole } = render(
				<Tooltip label="Interactive input tooltip">
					<TextInput
						name="test-input-2"
						placeholder="Test input"
						onChange={handleChange}
					/>
				</Tooltip>,
			);

			const input = getByRole('textbox');
			fireEvent.change(input, { target: { value: 'test' } });
			expect(handleChange).toHaveBeenCalled();
		});
	});

	describe('with native button element', () => {
		it('should work with native button', () => {
			const { container, getByRole } = render(
				<Tooltip label="Native button tooltip">
					<button type="button">Native Button</button>
				</Tooltip>,
			);

			const button = getByRole('button');
			expect(button).toHaveTextContent('Native Button');

			fireEvent.mouseEnter(button);
			expect(container.parentNode).toHaveTextContent(
				'Native button tooltip',
			);
		});

		it('should preserve native button functionality', () => {
			const handleClick = vi.fn();
			const { getByRole } = render(
				<Tooltip label="Native clickable tooltip">
					<button type="button" onClick={handleClick}>
						Native Clickable
					</button>
				</Tooltip>,
			);

			const button = getByRole('button');
			fireEvent.click(button);
			expect(handleClick).toHaveBeenCalledTimes(1);
		});
	});
});
