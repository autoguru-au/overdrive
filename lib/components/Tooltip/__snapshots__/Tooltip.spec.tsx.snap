// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Tooltip /> > should match snapshot when closed 1`] = `
<body
  class="theme_OD_Base__j49br80"
>
  <div>
    <div
      class="theme_OD_Base__j49br80"
      data-od-component="provider"
    >
      <span
        class="sprinkles_display_inline-block_mobile__onxwju7j"
        tabindex="0"
      >
        <div>
          trigger
        </div>
      </span>
    </div>
  </div>
  <div
    class="theme_OD_Base__j49br80"
  >
    <div
      aria-hidden="true"
      class="elementReset_resetVariants__18vhvlid elementReset_resetVariants_as_div__18vhvlie elementReset_block__18vhvli4 sprinkles_position_fixed_mobile__onxwju97 Positioner_root__1d54b7d0"
      data-od-component="positioner"
    />
  </div>
</body>
`;

exports[`<Tooltip /> > should match snapshot when opened 1`] = `
<body
  class="theme_OD_Base__j49br80"
>
  <div>
    <div
      class="theme_OD_Base__j49br80"
      data-od-component="provider"
    >
      <span
        aria-describedby="«r3»"
        class="sprinkles_display_inline-block_mobile__onxwju7j"
        tabindex="0"
      >
        <div>
          trigger
        </div>
      </span>
    </div>
  </div>
  <div
    class="theme_OD_Base__j49br80"
  >
    <div
      aria-hidden="false"
      class="elementReset_resetVariants__18vhvlid elementReset_resetVariants_as_div__18vhvlie elementReset_block__18vhvli4 sprinkles_position_fixed_mobile__onxwju97 Positioner_root__1d54b7d0"
      data-od-component="positioner"
      style="position: absolute; left: 0px; top: 0px; margin: 0px;"
    >
      <div
        class="elementReset_resetVariants__18vhvlid elementReset_resetVariants_as_div__18vhvlie elementReset_block__18vhvli4 sprinkles_paddingLeft_3_mobile__onxwjut3 sprinkles_paddingRight_3_mobile__onxwjuu7 sprinkles_paddingTop_2_mobile__onxwjuv7 sprinkles_paddingBottom_2_mobile__onxwjurv sprinkles_overflow_hidden_mobile__onxwju7z sprinkles_pointerEvents_none__onxwju6m sprinkles_userSelect_none__onxwju6p sprinkles_width_full_mobile__onxwjufn sprinkles_backgroundColour_gray900__onxwju4e sprinkles_boxShadow_4__onxwju6i sprinkles_borderRadius_1_mobile__onxwju9v Tooltip_root__1qliafn0"
        id="«r3»"
        role="tooltip"
      >
        <span
          class="elementReset_resetVariants__18vhvlid elementReset_resetVariants_as_span__18vhvlim  sprinkles_textAlign_left_mobile__onxwju9f typography_typographyStyles__yhjp331j typography_typographyStyles_as_span__yhjp331m  typography_typographyStyles_weight_normal__yhjp332o typography__yhjp331c sprinkles_fontWeight_normal__onxwju62 typography_typographyStyles_colour_white__yhjp3321 typography__yhjp33s typography_color_white__yhjp336 typography_typographyStyles_size_2__yhjp332c typography__yhjp3311 sprinkles_fontSize_2__onxwju5u sprinkles_lineHeight_2__onxwju5l"
        >
          tooltip content
        </span>
      </div>
    </div>
  </div>
</body>
`;
